# Введение начальных условий

Ниже введены следующие обозначения:
p - вероятности наступления спроса
count - количество мешков с цементом
save_price - цена за единицу при хранении
defic_price - цена за единицу в случае возникновления дефицита

In [1]:
p = [0.2, 0.5, 0.3]
count = [1500, 2000, 2500]
save_price = 5
defic_price = 3

# Добавление фукнции для красивой печати матриц

In [2]:
def printA(a):
    for row in a:
        for col in row:
            print("{:8.1f}".format(col), end=" ")
        print("")

# Составление таблицы
По итогу получается таблица, отражающая потери при различном объеме спроса

In [3]:
table = [[0 for i in range(len(p))] for j in range(len(count))]
for i in range(len(p)):
    for j in range(len(count)):
        demand = count[i] # объем спроса
        offer = count[j] # Объем предложения
        total_save_price = - (max(offer - demand, 0)) * save_price # затраты на хранение излишков
        total_defic_price = - (max(demand - offer, 0)) * defic_price # упущенная прибыль при дефиците
        table[i][j] = p[i] * (total_save_price + total_defic_price) # общие потери

printA(table)

     0.0   -500.0  -1000.0 
  -750.0      0.0  -1250.0 
  -900.0   -450.0      0.0 


# Принятие решения на основании Критерий максимакса.

In [4]:
maximum_criterion = [max(row) for row in table]
maximum_criterion

[0.0, 0.0, 0.0]

## На основании него невозможно определить выигрышную стратегию

# Принятие решение на основании Максиминный критерий Вальда.

In [5]:
maximin_wald_test = [min(row) for row in table]
maximin_wald_test

[-1000.0, -1250.0, -900.0]

## На основании него выигрышной стратегией является закупка 2500 т.к. наибольший среди наименьших выигрышей появляется в 3 случае и равен -900

# Принятие решение на основании Критерий минимаксного риска Сэвиджа.

In [6]:
b = [max(row) for row in table]
sub_table = [[b[j] - table[i][j] for j in range(len(table))] for i in range(len(table))]
print('Матрица рисков')
printA(sub_table)
savage_minimax_risk_criterion = [max(row) for row in sub_table]
savage_minimax_risk_criterion

Матрица рисков
     0.0    500.0   1000.0 
   750.0      0.0   1250.0 
   900.0    450.0      0.0 


[1000.0, 1250.0, 900.0]

## На основании него выигрышной стратегией является закупка 2500 т.к. при ней достигается наименьший уровень риска 900

# Принятие решения на основании Критерий пессимизма-оптимизма Гурвица.

In [7]:
hurwitz_pessimism_optimism_criterion = [0.5 * min(row) + 0.5 * max(row) for row in table]
hurwitz_pessimism_optimism_criterion

[-500.0, -625.0, -450.0]

## На основании него выигрышной стратегией является закупка 2000 т.к. при этом достигается наименьшее значение -625

# Принятие решение на основании риска

In [8]:
result = [sum(row) for row in table]
result

[-1500.0, -2000.0, -1350.0]

## На основании этого можно говорить, что выигрышная стратегия - третья

# Ввод начальных значений
Ниже введены следующие обозначения:
people_stream - поток людей
price - стоимость услуги
lost_men_price - потери, связаные с потерей посетителя
prostoi_price - потери, связаные с простоем работника

In [9]:
people_stream = [2000, 4000, 6000, 8000]
price = 90
lost_men_price = 50
prostoi_price = 60

# Генерация таблицы

In [10]:
table = [[0 for i in range(len(people_stream))] for j in range(len(people_stream))]
for i in range(len(table)):
    for j in range(len(table[i])):
        prishlo = people_stream[j] # количество людей, пришедших на ремонт
        dumali = people_stream[i] # ожидаемый поток клиентов
        lost = max(prishlo - dumali, 0) # количество людей, которых мы не смогли обслужить
        prostoi = max(dumali - prishlo, 0) # количество простоя
        viruchka = min(dumali, prishlo) * price # выручка
        poteri = lost * lost_men_price + prostoi * prostoi_price # потери от необслуженных клиентов и простоя
        table[i][j] = viruchka - poteri # прибыль

printA(table)

180000.0  80000.0 -20000.0 -120000.0 
 60000.0 360000.0 260000.0 160000.0 
-60000.0 240000.0 540000.0 440000.0 
-180000.0 120000.0 420000.0 720000.0 


# Принятие решения на основании Критерий максимакса.

In [11]:
maximum_criterion = [max(row) for row in table]
maximum_criterion

[180000, 360000, 540000, 720000]

## На основании него выигрышной стратегией является закупка 8000, т.к. максимальный выигрыш достигается в 4 случае и равен 720000

# Принятие решение на основании Максиминный критерий Вальда.

In [12]:
maximin_wald_test = [min(row) for row in table]
maximin_wald_test

[-120000, 60000, -60000, -180000]

## На основании него выигрышной стратегией является закупка 4000 т.к. наибольший среди наименьших выигрышей появляется во 2 случае и равен 60000

# Принятие решение на основании Критерий минимаксного риска Сэвиджа.

In [13]:
b = [max(row) for row in table]
sub_table = [[b[j] - table[i][j] for j in range(len(table))] for i in range(len(table))]
print('Матрица рисков')
printA(sub_table)
savage_minimax_risk_criterion = [max(row) for row in sub_table]
savage_minimax_risk_criterion

Матрица рисков
     0.0 280000.0 560000.0 840000.0 
120000.0      0.0 280000.0 560000.0 
240000.0 120000.0      0.0 280000.0 
360000.0 240000.0 120000.0      0.0 


[840000, 560000, 280000, 360000]

## На основании него выигрышной стратегией является закупка 6000 т.к. при ней достигается наименьший уровень риска 280000

# Принятие решения на основании Критерий пессимизма-оптимизма Гурвица.

In [14]:
hurwitz_pessimism_optimism_criterion = [0.5 * min(row) + 0.5 * max(row) for row in table]
hurwitz_pessimism_optimism_criterion

[30000.0, 210000.0, 240000.0, 270000.0]

## На основании него выигрышной стратегией является закупка 2000 т.к. при этом достигается наименьшее значение 30000